core/src/main/java/com/linecorp/armeria/internal/common/util/TransportTypeProvider.java

```diff
@@ -0,0 +1,213 @@
+ * under the License.
+ */
+package com.linecorp.armeria.internal.common.util;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.util.concurrent.ThreadFactory;
+import java.util.function.BiFunction;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.common.util.Exceptions;
+import com.linecorp.armeria.common.util.TransportType;
+
+import io.netty.channel.EventLoop;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.nio.NioEventLoop;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.DatagramChannel;
+import io.netty.channel.socket.ServerSocketChannel;
+import io.netty.channel.socket.SocketChannel;
+import io.netty.channel.socket.nio.NioDatagramChannel;
+import io.netty.channel.socket.nio.NioServerSocketChannel;
+import io.netty.channel.socket.nio.NioSocketChannel;
+
+/**
+ * Provides the properties required by {@link TransportType} by loading /dev/epoll and io_uring transport
+ * classes dynamically, so that Armeria does not have hard dependencies on them.
+ * See: https://github.com/line/armeria/issues/3243
+ */
+public final class TransportTypeProvider {
+
+    public static final TransportTypeProvider NIO = new TransportTypeProvider(
+            "NIO", NioServerSocketChannel.class, NioSocketChannel.class, NioDatagramChannel.class,
+            NioEventLoopGroup.class, NioEventLoop.class, NioEventLoopGroup::new, null);
+
+    public static final TransportTypeProvider EPOLL = of(
+            "EPOLL", "io.netty.channel.epoll.Epoll",
+            "io.netty.channel.epoll.EpollServerSocketChannel",
+            "io.netty.channel.epoll.EpollSocketChannel",
+            "io.netty.channel.epoll.EpollDatagramChannel",
+            "io.netty.channel.epoll.EpollEventLoopGroup",
+            "io.netty.channel.epoll.EpollEventLoop");
+
+    public static final TransportTypeProvider IO_URING = of(
+            "IO_URING", "io.netty.incubator.channel.uring.IOUring",
+            "io.netty.incubator.channel.uring.IOUringServerSocketChannel",
+            "io.netty.incubator.channel.uring.IOUringSocketChannel",
+            "io.netty.incubator.channel.uring.IOUringDatagramChannel",
+            "io.netty.incubator.channel.uring.IOUringEventLoopGroup",
+            "io.netty.incubator.channel.uring.IOUringEventLoop");
+
+    private static TransportTypeProvider of(
+            String name, String entryPointTypeName,
+            String serverSocketChannelTypeName, String socketChannelTypeName, String datagramChannelTypeName,
+            String eventLoopGroupTypeName, String eventLoopTypeName) {
+
+        try {
+            final Throwable unavailabilityCause = (Throwable)
+                    findClass(entryPointTypeName)
+                            .getMethod("unavailabilityCause")
+                            .invoke(null);
+            if (unavailabilityCause != null) {
+                throw unavailabilityCause;
```

Comment:
Shouldn't we create `TransportTypeProvider` with the cause? 🤔

Reply:
We catch below 😉

Reply:
Oops. 😅